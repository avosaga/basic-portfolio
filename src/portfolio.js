import { isDate } from './utils.js'

export class Portfolio {
    #stocks = []

    add(stock) {
        this.#stocks.push(stock)
        return this
    }

    async profit (from, to) {
        if (!isDate(from) || !isDate(to)) {
            throw new Error('date range not provided, object { from, to } has to be defined')
        }

        const annualizedReturn = []
        let bookValue = 0
        let bookProfit = 0

        for (const stock of this.#stocks) {
            if (stock.purchaseDate >= from && stock.purchaseDate <= to) {
                const quantity = stock.quantity
                const currentPrice = await stock.getPrice(to)
                const purchasePrice = await stock.getPrice()
                const profit = currentPrice - purchasePrice

                annualizedReturn.push({
                    symbol: stock.symbol,
                    currentPrice,
                    purchasePrice,
                    profit,
                    quantity
                })

                bookValue += purchasePrice * quantity
                bookProfit += profit * quantity
            }
        }

        return {
            bookValue,
            bookProfit,
            annualizedReturn
        }
    }
}
