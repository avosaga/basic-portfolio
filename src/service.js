import { isEmptyString, isDate } from './utils.js'

const symbolMap = new Map()

export function getPriceBySymbol (symbol = '', date = new Date()) {
    return new Promise((resolve, reject) => {
        if (isEmptyString(symbol)) {
            reject(new Error('symbol is empty'))
        }

        if (!isDate(date)) {
            reject(new Error('date is invalid, must be defined and instance of Date'))
        }

        const datePriceKey = date.toDateString()
        const datePriceMap = symbolMap.has(symbol) ? symbolMap.get(symbol) : new Map()

        let price
        if (datePriceMap.size > 0 && datePriceMap.has(datePriceKey)) {
            price = datePriceMap.get(datePriceKey)
        } else {
            price = Math.random() * 100
            datePriceMap.set(datePriceKey, price)
            symbolMap.set(symbol, datePriceMap)
        }

        resolve(price)
    })
}
