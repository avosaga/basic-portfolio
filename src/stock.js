import { getPriceBySymbol } from './service.js'

export class Stock {
    #symbol
    #purchaseDate
    #quantity

    constructor (symbol, purchaseDate = new Date(), quantity = 1) {
        this.#symbol = symbol
        this.#purchaseDate = purchaseDate
        this.#quantity = quantity
    }

    get purchaseDate () {
        return this.#purchaseDate
    }

    get quantity () {
        return this.#quantity
    }

    get symbol () {
        return this.#symbol
    }

    async getPrice (date = this.#purchaseDate) {
        return await getPriceBySymbol(this.#symbol, date)
    }
}
