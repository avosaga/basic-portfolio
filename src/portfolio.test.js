import { Portfolio } from './portfolio'
import { Stock } from './stock'
import { getPriceBySymbol } from './service'

jest.mock('./service', () => ({
    getPriceBySymbol: jest.fn()
}))

describe('Portfolio', () => {
    it('should throw error if date range not provided', async () => {
        const portfolio = new Portfolio()

        await expect(
            portfolio.profit()
        ).rejects.toThrowError('date range not provided, object { from, to } has to be defined')
    })

    it('should return default values when no stocks are defined', async () => {
        const portfolio = new Portfolio()

        const { bookValue, annualizedReturn, bookProfit } = await portfolio.profit(
            new Date(2022, 0, 1),
            new Date(2022, 5, 1)
        )

        expect(bookValue).toBe(0)
        expect(bookProfit).toBe(0)
        expect(annualizedReturn).toEqual([])
    })

    it('should return calculated values when stocks are provided', async () => {
        const google = 'GOOGLE'
        const apple = 'APPLE'
        const tesla = 'TESLA'
        const googlePrice = 1
        const applePrice = 2
        const teslaPrice = 3

        getPriceBySymbol.mockImplementation((symbol) => {
            switch (symbol) {
            case google: return Promise.resolve(googlePrice)
            case apple: return Promise.resolve(applePrice)
            case tesla: return Promise.resolve(teslaPrice)
            }
        })
        const portfolio = new Portfolio()
        portfolio
            .add(new Stock(google, new Date(2022, 1, 1)))
            .add(new Stock(apple, new Date(2022, 2, 2)))
            .add(new Stock(tesla, new Date(2021, 11, 1)))

        const { bookValue, annualizedReturn, bookProfit } = await portfolio.profit(
            new Date(2021, 0, 1),
            new Date(2022, 5, 1)
        )

        expect(bookValue).toEqual(googlePrice + applePrice + teslaPrice)
        expect(bookProfit).toEqual(0)
        expect(annualizedReturn).toMatchSnapshot()
    })
})
