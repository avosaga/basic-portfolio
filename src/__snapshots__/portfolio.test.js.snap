// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio should return calculated values when stocks are provided 1`] = `
Array [
  Object {
    "currentPrice": 1,
    "profit": 0,
    "purchasePrice": 1,
    "quantity": 1,
    "symbol": "GOOGLE",
  },
  Object {
    "currentPrice": 2,
    "profit": 0,
    "purchasePrice": 2,
    "quantity": 1,
    "symbol": "APPLE",
  },
  Object {
    "currentPrice": 3,
    "profit": 0,
    "purchasePrice": 3,
    "quantity": 1,
    "symbol": "TESLA",
  },
]
`;
